lib/matplotlib/backends/qt_compat.py

```diff
@@ -1,220 +1,159 @@
-""" A Qt API selector that can be used to switch between PyQt and PySide.
+"""
+Qt binding and backend selector.
+
+The selection logic is as follows:
+- if any of PyQt5, PySide2, PyQt4 or PySide have already been imported
+  (checked in that order), use it;
+- otherwise, if the QT_API environment variable (used by Enthought) is
+  set, use it to determine which binding to use (but do not change the
+  backend based on it; i.e. if the Qt4Agg backend is requested but QT_API
+  is set to "pyqt5", then actually use Qt4 with the binding specified by
+  ``rcParams["backend.qt4"]``;
+- otherwise, use whatever the rcParams indicate.
 """
 
+from distutils.version import LooseVersion
 import os
-import logging
 import sys
-from matplotlib import rcParams
 
-_log = logging.getLogger(__name__)
+from matplotlib import rcParams
 
-# Available APIs.
-QT_API_PYQT = 'PyQt4'       # API is not set here; Python 2.x default is V 1
-QT_API_PYQTv2 = 'PyQt4v2'   # forced to Version 2 API
-QT_API_PYSIDE = 'PySide'    # only supports Version 2 API
-QT_API_PYQT5 = 'PyQt5'      # use PyQt5 API; Version 2 with module shim
-QT_API_PYSIDE2 = 'PySide2'  # Version 2 API with module shim
-
-ETS = dict(pyqt=(QT_API_PYQTv2, 4), pyside=(QT_API_PYSIDE, 4),
-           pyqt5=(QT_API_PYQT5, 5), pyside2=(QT_API_PYSIDE2, 5))
-# ETS is a dict of env variable to (QT_API, QT_MAJOR_VERSION)
-# If the ETS QT_API environment variable is set, use it, but only
-# if the variable is of the same major QT version.  Note that
-# ETS requires the version 2 of PyQt4, which is not the platform
-# default for Python 2.x.
 
+QT_API_PYQT5 = "PyQt5"
+QT_API_PYSIDE2 = "PySide2"
+QT_API_PYQTv2 = "PyQt4v2"
+QT_API_PYSIDE = "PySide"
+QT_API_PYQT = "PyQt4"
 QT_API_ENV = os.environ.get('QT_API')
-
-if rcParams['backend'] == 'Qt5Agg':
-    QT_RC_MAJOR_VERSION = 5
-elif rcParams['backend'] == 'Qt4Agg':
-    QT_RC_MAJOR_VERSION = 4
-else:
-    # A different backend was specified, but we still got here because a Qt
-    # related file was imported. This is allowed, so lets try and guess
-    # what we should be using.
-    if "PyQt4" in sys.modules or "PySide" in sys.modules:
-        # PyQt4 or PySide is actually used.
-        QT_RC_MAJOR_VERSION = 4
-    else:
-        # This is a fallback: PyQt5
-        QT_RC_MAJOR_VERSION = 5
-
-QT_API = None
-
-# check if any binding is already imported, if so silently ignore the
-# rcparams/ENV settings and use what ever is already imported.
-if 'PySide' in sys.modules:
-    # user has imported PySide before importing mpl
-    QT_API = QT_API_PYSIDE
-
-if 'PySide2' in sys.modules:
-    # user has imported PySide before importing mpl
+# First, check if anything is already imported.
```

Comment:
I assume we are prepared to do quick patch releases if this stuff doesn't work on everyone's setup....